macOS/debugger.cpp

```diff
@@ -68,12 +68,37 @@ vm_prot_t Debugger::MacOSProtectionFlags(MemoryProtection memory_protection) {
   }
 }
 
+void Debugger::FreeShare(void *address, size_t size) {
+  if (size == 0) {
+    WARN("FreeShare is called with size == 0\n");
+    return;
+  }
+
+  mach_port_t shm_port = share_mem_map[(mach_vm_address_t)address];
+  kern_return_t krt = mach_port_destroy(mach_target->Task(), shm_port);
+  if (krt != KERN_SUCCESS) {
+    FATAL("Error (%s) destroy port for shared memory @ 0x%llx\n", mach_error_string(krt), (mach_vm_address_t)address);
+  }
+
+  krt = mach_vm_deallocate(mach_task_self(), (mach_vm_address_t)address, (mach_vm_size_t)size);
+  if (krt != KERN_SUCCESS) {
+    FATAL("Error (%s) freeing memory @ 0x%llx\n", mach_error_string(krt), (mach_vm_address_t)address);
+  }
+}
+
 void Debugger::RemoteFree(void *address, size_t size) {
+  if (share_mem_map.find((mach_vm_address_t)address) != share_mem_map.end()) {
+    FreeShare((void *)share_mem_map[(mach_vm_address_t)address], size);
+  }
   mach_target->FreeMemory((uint64_t)address, size);
 }
 
 void Debugger::RemoteRead(void *address, void *buffer, size_t size) {
-  mach_target->ReadMemory((uint64_t)address, size, buffer);
+  if (share_mem_map.find((mach_vm_address_t)address) != share_mem_map.end()) {
```

Comment:
This will only work when reading from the very start address of the shared memory. I suggest saving both the address and the size in the map, and checking if the range to be read falls into the address range of any of the shared memory objects. Note that walking the entire list of shared memory objects here will still be faster than cross-process read memory.